borg/hashindex.pyx

```diff
@@ -175,9 +192,39 @@ cdef class ChunkIndex(IndexBase):
 
     def __contains__(self, key):
         assert len(key) == self.key_size
-        data = <int *>hashindex_get(self.index, <char *>key)
+        data = <int32_t *>hashindex_get(self.index, <char *>key)
         return data != NULL
 
+    def incref(self, key):
+        """Increase refcount for 'key', return (refcount, size, csize)"""
+        assert len(key) == self.key_size
+        data = <int32_t *>hashindex_get(self.index, <char *>key)
+        if not data:
+            raise KeyError(key)
+        cdef int32_t refcount = _le32toh(data[0])
+        if refcount != INT32_MAX:
+            if INT32_MAX - refcount < 1:
```

Comment:
@verygreen Ok, I see what you mean, but `INT32_MAX - -1` is the undefined behaviour we want to avoid here. I'm not sure we should consider negative refcounts; if the refcount already went beyond 2**31 then we don't really know how far it went. It could already crossed over zero again (with data loss). (For 1.x the issue is a bit academic, because the only kind of data that would realistically make very high refcounts are large [possibly sparse] disk images with lots of zeroes, but an all zero chunk would always go to MAX_EXP in the chunker [my understanding], so 2**31-1 * 8 MB ~ 20 PB)